message_sender/tasks.py

```diff
@@ -77,6 +80,71 @@ def run(self, metric_name, metric_value, session=None, **kwargs):
 fire_metric = FireMetric()
 
 
+class ConcurrencyLimiter(object):
+    BUCKET_SIZE = 60
+
+    @classmethod
+    def get_current_message_count(cls, msg_type, delay):
+        # Sum the values in all the buckets to get the total
+        total = 0
+        number_of_buckets = delay // cls.BUCKET_SIZE + 1
+        bucket = int(time.time() // cls.BUCKET_SIZE)
+        for i in range(0, number_of_buckets):
+            value = cache.get(msg_type+"_messages_at_%s" % bucket)
+            bucket = bucket - 1
+            if value:
+                total += int(value)
+        return total
+
+    @classmethod
+    def incr_message_count(cls, msg_type, delay):
+        bucket = int(time.time() // cls.BUCKET_SIZE)
+        key = msg_type + "_messages_at_%s" % bucket
+        value = cache.get(key)
+        if value is None:
+            # Add the bucket size to the expiry time so messages that start at
+            # the end of the bucket still complete
+            cache.set(key, 1, delay + cls.BUCKET_SIZE)
+        else:
+            cache.incr(key)
+
+    @classmethod
+    def decr_message_count(cls, msg_type, msg_time):
+        if msg_type == "voice":
+            delay = getattr(settings, 'VOICE_MESSAGE_DELAY', 0)
+        else:
+            delay = getattr(settings, 'TEXT_MESSAGE_DELAY', 0)
+
+        if not msg_time:
+            return
+
+        # Convert from datetime to seconds since epoch
+        msg_time = (msg_time - datetime(1970, 1, 1)).total_seconds()
+
+        time_since = time.time() - msg_time
+        if time_since > delay:
+            return
+        bucket = int(msg_time // cls.BUCKET_SIZE)
+
+        key = msg_type + "_messages_at_%s" % bucket
+        value = cache.get(key)
+        # Don't allow negative values
+        if value:
+            if value < 0:
+                # Set the expiry time to the delay minus the time passed since
+                # the message was sent.
+                cache.set(key, 0, delay - time_since)
+            else:
+                cache.decr(key)
+
+    @classmethod
+    def manage_limit(cls, task, msg_type, limit, delay):
+        if limit > 0:
+            if cls.get_current_message_count(msg_type, delay) >= limit:
+                task.retry(countdown=delay)
+            cls.incr_message_count(msg_type, delay)
+
+
 class Send_Message(Task):
```

Comment:
I'm not sure if this is set, but we might want to set the max retries on this task to None, so that we don't run out of retries if we have a lot of messages in the queue with not so many slots: http://docs.celeryproject.org/en/latest/reference/celery.app.task.html#celery.app.task.Task.max_retries